File: src/packages/sdk-python/src/evalai_sdk/regression.py
"""Regression gate constants, types, and helpers."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal


class GATE_EXIT:
    PASS = 0
    REGRESSION = 1
    INFRA_ERROR = 2
    CONFIDENCE_FAILED = 3
    CONFIDENCE_MISSING = 4


class GATE_CATEGORY:
    PASS = "pass"
    REGRESSION = "regression"
    INFRA_ERROR = "infra_error"


REPORT_SCHEMA_VERSION = 1


class ARTIFACTS:
    BASELINE = "evals/baseline.json"
    REGRESSION_REPORT = "evals/regression-report.json"
    CONFIDENCE_SUMMARY = "evals/confidence-summary.json"
    LATENCY_BENCHMARK = "evals/latency-benchmark.json"


@dataclass
class BaselineTolerance:
    score_drop: float = 0.05
    latency_increase_pct: float = 20.0
    min_confidence: float = 0.8


@dataclass
class Baseline:
    version: int = REPORT_SCHEMA_VERSION
    scores: dict[str, float] = field(default_factory=dict)
    latencies: dict[str, float] = field(default_factory=dict)
    created_at: str | None = None
    tolerance: BaselineTolerance = field(default_factory=BaselineTolerance)
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class RegressionDelta:
    test_id: str
    metric: str
    baseline_value: float
    current_value: float
    delta: float
    delta_pct: float
    category: str = GATE_CATEGORY.PASS
    severity: Literal["low", "medium", "high", "critical"] = "low"


@dataclass
class RegressionReport:
    version: int = REPORT_SCHEMA_VERSION
    run_id: str = ""
    gate_exit: int = GATE_EXIT.PASS
    gate_category: str = GATE_CATEGORY.PASS
    deltas: list[RegressionDelta] = field(default_factory=list)
    summary: dict[str, Any] = field(default_factory=dict)
    created_at: str | None = None


def evaluate_regression(
    baseline: Baseline,
    current_scores: dict[str, float],
    *,
    min_score: float | None = None,
    max_drop: float | None = None,
) -> RegressionReport:
    """Compare current scores against a baseline and produce a regression report.

    Args:
        baseline: Baseline scores and tolerance configuration.
        current_scores: Current run scores keyed by test ID.
        min_score: Optional absolute floor — any score below this is a failure.
        max_drop: Optional override for baseline.tolerance.score_drop.
    """
    effective_drop = max_drop if max_drop is not None else baseline.tolerance.score_drop

    deltas: list[RegressionDelta] = []
    gate_exit = GATE_EXIT.PASS
    gate_category = GATE_CATEGORY.PASS

    for test_id, baseline_score in baseline.scores.items():
        current = current_scores.get(test_id)
        if current is None:
            continue
        delta = current - baseline_score
        delta_pct = (delta / baseline_score * 100) if baseline_score != 0 else 0

        failed = False

        if delta < -effective_drop:
            failed = True

        if min_score is not None and current < min_score:
            failed = True

        if failed:
            severity: Literal["low", "medium", "high", "critical"] = "high" if abs(delta_pct) > 20 else "medium"
            deltas.append(
                RegressionDelta(
                    test_id=test_id,
                    metric="score",
                    baseline_value=baseline_score,
                    current_value=current,
                    delta=delta,
                    delta_pct=delta_pct,
                    category=GATE_CATEGORY.REGRESSION,
                    severity=severity,
                )
            )
            gate_exit = GATE_EXIT.REGRESSION
            gate_category = GATE_CATEGORY.REGRESSION
        else:
            deltas.append(
                RegressionDelta(
                    test_id=test_id,
                    metric="score",
                    baseline_value=baseline_score,
                    current_value=current,
                    delta=delta,
                    delta_pct=delta_pct,
                )
            )

    # Check min_score for tests not in the baseline
    if min_score is not None:
        for test_id, current in current_scores.items():
            if test_id in baseline.scores:
                continue
            if current < min_score:
                deltas.append(
                    RegressionDelta(
                        test_id=test_id,
                        metric="score",
                        baseline_value=0.0,
                        current_value=current,
                        delta=current,
                        delta_pct=0.0,
                        category=GATE_CATEGORY.REGRESSION,
                        severity="medium",
                    )
                )
                gate_exit = GATE_EXIT.REGRESSION
                gate_category = GATE_CATEGORY.REGRESSION

    return RegressionReport(
        gate_exit=gate_exit,
        gate_category=gate_category,
        deltas=deltas,
        summary={
            "total": len(deltas),
            "regressions": sum(1 for d in deltas if d.category == GATE_CATEGORY.REGRESSION),
            "passed": sum(1 for d in deltas if d.category == GATE_CATEGORY.PASS),
        },
    )

File: src/packages/sdk/src/regression.ts
/**
 * Regression gate constants and types.
 *
 * These mirror the contracts defined in scripts/regression-gate.ts
 * and evals/schemas/regression-report.schema.json so that SDK consumers
 * can programmatically inspect gate results without parsing strings.
 *
 * @packageDocumentation
 */

// ── Exit codes ──

/** Exit codes emitted by `evalai gate` / `scripts/regression-gate.ts`. */
export const GATE_EXIT = {
	/** Gate passed — no regressions detected */
	PASS: 0,
	/** One or more regression thresholds exceeded */
	REGRESSION: 1,
	/** Infrastructure error (baseline missing, summary missing, etc.) */
	INFRA_ERROR: 2,
	/** Confidence tests failed (test suite red) */
	CONFIDENCE_FAILED: 3,
	/** Confidence summary file missing (test infra crashed) */
	CONFIDENCE_MISSING: 4,
} as const;

export type GateExitCode = (typeof GATE_EXIT)[keyof typeof GATE_EXIT];

// ── Report categories ──

/** Categories written to regression-report.json `category` field. */
export const GATE_CATEGORY = {
	PASS: "pass",
	REGRESSION: "regression",
	INFRA_ERROR: "infra_error",
} as const;

export type GateCategory = (typeof GATE_CATEGORY)[keyof typeof GATE_CATEGORY];

// ── Schema version ──

/** Current schema version for regression-report.json. */
export const REPORT_SCHEMA_VERSION = 1;

// ── Report types ──

export interface RegressionDelta {
	metric: string;
	baseline: number | string;
	current: number | string;
	delta: string;
	status: "pass" | "fail";
}

export interface RegressionReport {
	schemaVersion: number;
	timestamp: string;
	exitCode: GateExitCode;
	category: GateCategory;
	passed: boolean;
	failures: string[];
	deltas: RegressionDelta[];
}

// ── Baseline types ──

export interface BaselineTolerance {
	scoreDrop: number;
	passRateDrop: number;
	maxLatencyIncreaseMs: number;
	maxCostIncreaseUsd: number;
}

export interface Baseline {
	schemaVersion: number;
	description: string;
	generatedAt: string;
	generatedBy: string;
	commitSha: string;
	updatedAt: string;
	updatedBy: string;
	tolerance: BaselineTolerance;
	goldenEval: {
		score: number;
		passRate: number;
		totalCases: number;
		passedCases: number;
	};
	qualityScore: {
		overall: number;
		grade: string;
		accuracy: number;
		safety: number;
		latency: number;
		cost: number;
		consistency: number;
	};
	confidenceTests: {
		unitPassed: boolean;
		unitTotal: number;
		dbPassed: boolean;
		dbTotal: number;
	};
	productMetrics: {
		p95ApiLatencyMs?: number;
		goldenCostUsd?: number;
	};
	qualityMetrics?: {
		unitLaneDurationMs?: number;
		dbLaneDurationMs?: number;
	};
}

// ── Artifact paths ──

/** Well-known artifact paths relative to project root. */
export const ARTIFACTS = {
	BASELINE: "evals/baseline.json",
	REGRESSION_REPORT: "evals/regression-report.json",
	CONFIDENCE_SUMMARY: "evals/confidence-summary.json",
	LATENCY_BENCHMARK: "evals/latency-benchmark.json",
} as const;
